plutus-tx: tracing support

michaelpj opened the pull request:
- Add a `trace :: Builtins.String -> ()` function to the plugin builtins, mapped to the builtin trace.
- Add some TH trace functions more like those in `Debug.Trace` to `plutus-tx` to help with getting evaluation order right.
- Add some evaluation functions to `plutus-tx` that hide supplying the dynamic builtins (and also, sigh, more `unsafePerformIO` that I think we need).
- Use the evaluation functions in `wallet-api` and include the log output in validation failure errors. AFAICT there's currently nothing that tests the log events, so we should add such a tests and check that it includes trace output for failed transactions.

effectfully commented on plutus-tx-plugin/src/Language/PlutusTx/Compiler/Builtins.hs:
```diff
@@ -260,13 +264,16 @@ defineBuiltinTerms = do
     do
         let term = mkDynBuiltin PLC.dynamicAppendName
-        defineBuiltinTerm 'Builtins.appendString term []
+        defineBuiltinTerm 'Builtins.appendString term [str]
```
What does `str` mean here?

michaelpj replied:
Oh, this is just something I should have done before. These are dependencies between definitions so that PIR can generate the lets in the right order. In this case it doesn't matter much but it's just hygienic to put the dependencies in. (`appendString` needs the `string` type!).

effectfully commented on plutus-tx/src/Language/PlutusTx/Evaluation.hs:
```diff
@@ -0,0 +1,34 @@
+evaluateCek
+    :: Program TyName Name ()
+    -> EvaluationResult
+evaluateCek p = unsafePerformIO $ evaluate $ runCek stringBuiltins p
```
Is `unsafePerformIO . evaluate` not the same thing as `id`?

michaelpj replied:
Honestly I'm not sure. Is it? `evaluate` does some forcing.

mchakravarty replied:
`evaluate` force evaluation to WHNF, but that'll happen anyway when this expression is evaluated; so, yes, it is `== id`.

effectfully commented on plutus-tx/src/Language/PlutusTx/Evaluation.hs:
```diff
@@ -0,0 +1,34 @@
+
+-- | Evaluate a program in the CEK machine with the usual string dynamic builtins and tracing, additionally
+-- returning the trace output.
+evaluateCekLog
```
I think it's better then to use `trace` consistently everywhere rather than `log`.

michaelpj replied:
:+1:

effectfully commented on plutus-tx/src/Language/PlutusTx/Evaluation.hs:
```diff
@@ -0,0 +1,34 @@
+-- returning the trace output.
+evaluateCekLog
+    :: Program TyName Name ()
+    -> ([String], EvaluationResult)
```
We could keep this in the honest `IO` at the cost of complicating the golden test which I think is fine, but that's not too important I guess.

michaelpj replied:
It's not the test that I'm worried about, it's all the wallet-api stuff, which is very annoying if this stops being pure.

michaelpj commented on wallet-api/src/Wallet/UTXO/Index.hs:
```diff
@@ -172,10 +172,11 @@ checkMatch v = \case
             in
-                if UTXO.runScript v' vl r d
+                if success
                 then pure ()
```
Yeah, it's a little awkward because we don't currently return anything in the success case, so it wasn't totally clear to me how to wire that through in the most appropriate way, whereas sticking it into the validation errors is obvious.

j-mueller replied:
I'll look into it when this is merged

michaelpj commented:
Okay, let me see if I can write a trace that has an invalid script.

michaelpj commented:
Working on getting an test working using `wallet-api`, which has thrown up that I'm not typechecking properly (which I hadn't noticed because I have to turn off typechecking for more-or-less everything that uses this, since it involves lists).

michaelpj commented:
Okay, I've fixed the typechecking (could do with a look from @effectfully, it's a bit awkward because we don't have the definition of `trace` at typechecking time), and added a test that has a failing script validation with a log, and verifies that it appears in the log events (could do with a look from @j-mueller ).

michaelpj commented:
(I was closer to done than I thought)

mchakravarty commented:
> "I always fail everything"

Sorry to hear that 😉

michaelpj commented:
Okay, let's get this in.
